File: snalib-dist/snalib-node/src/logger.ts
export class SnaLibLogger {
  private readonly logEnabled: boolean;
  private readonly onWarning: ((msg: string) => void) | undefined;

  constructor(logEnabled: boolean, onWarning?: (msg: string) => void) {
    this.logEnabled = logEnabled;
    this.onWarning = onWarning;
  }

  info(msg: string, ...kv: unknown[]): void {
    if (!this.logEnabled) return;
    process.stderr.write(this.buildJson('info', msg, kv) + '\n');
  }

  warn(msg: string, ...kv: unknown[]): void {
    const json = this.buildJson('warn', msg, kv);
    if (this.onWarning !== undefined) {
      // Pass the full JSON entry — same string that would otherwise go to stderr.
      this.onWarning(json);
      return;
    }
    process.stderr.write(json + '\n');
  }

  private buildJson(level: string, msg: string, kv: unknown[]): string {
    const entry: Record<string, unknown> = { level, msg };
    for (let i = 0; i + 1 < kv.length; i += 2) {
      const key = String(kv[i]);
      entry[key] = kv[i + 1];
    }
    return JSON.stringify(entry);
  }
}

File: snalib-dist/snalib-dotnet/src/Logger.cs
using System;
using System.Collections.Generic;
using System.Text.Json;

namespace Sahamati.SnaLib
{
    public class SnaLibLogger
    {
        private readonly bool _logEnabled;
        private readonly Action<string>? _onWarning;

        public SnaLibLogger(bool logEnabled, Action<string>? onWarning = null)
        {
            _logEnabled = logEnabled;
            _onWarning = onWarning;
        }

        public void Info(string msg, params object?[] kv)
        {
            if (!_logEnabled) return;
            Write(BuildJson("info", msg, kv));
        }

        public void Warn(string msg, params object?[] kv)
        {
            var json = BuildJson("warn", msg, kv);

            if (_onWarning != null)
            {
                try
                {
                    _onWarning(json);
                }
                catch
                {
                    // A caller callback that throws must not propagate. Warn() is called from
                    // inside the catch blocks of the public API methods, so an exception here
                    // would escape the SDK and fault the caller's task — breaking the
                    // never-throw contract (CLAUDE.md section 10). Fall back to stderr so the
                    // warning is still surfaced rather than swallowed.
                    Write(json);
                }
                return;
            }

            Write(json);
        }

        private static void Write(string line)
        {
            try
            {
                Console.Error.WriteLine(line);
            }
            catch
            {
                // stderr closed or redirected. Logging must never break the caller.
            }
        }

        private static string BuildJson(string level, string msg, object?[] kv)
        {
            var entry = new Dictionary<string, object?> { ["level"] = level, ["msg"] = msg };
            for (var i = 0; i + 1 < kv.Length; i += 2)
            {
                var key = kv[i]?.ToString();
                if (string.IsNullOrEmpty(key)) continue;
                entry[key!] = kv[i + 1];
            }

            try
            {
                return JsonSerializer.Serialize(entry);
            }
            catch
            {
                // A value that will not serialise must not cost us the log line.
                return JsonSerializer.Serialize(new Dictionary<string, object?>
                {
                    ["level"] = level,
                    ["msg"] = msg,
                });
            }
        }
    }
}
